src/eigen.jl

```diff
@@ -221,188 +221,160 @@ end
-    else
-        orthogonal1 = SVector(zero(S), eigvec1[3], -eigvec1[2]) / sqrt(abs2(eigvec1[2]) + abs2(eigvec1[3]))
+    function _sortperm3(v)
+        perm = SVector(1,2,3)
+        # unrolled bubble-sort
+        (v[perm[1]] > v[perm[2]]) && (perm = SVector(perm[2], perm[1], perm[3]))
+        (v[perm[2]] > v[perm[3]]) && (perm = SVector(perm[1], perm[3], perm[2]))
+        (v[perm[1]] > v[perm[2]]) && (perm = SVector(perm[2], perm[1], perm[3]))
+        perm
     end
-    orthogonal2 = eigvec1 × orthogonal1
-
-    # The projected 2x2 eigenvalue problem is C x = 0 where C is the projection
-    # of (A - eig2*I) onto the subspace {orthogonal1, orthogonal2}
-    @inbounds a_orth1_1 = a11 * orthogonal1[1] + a12 * orthogonal1[2] + a13 * orthogonal1[3]
-    @inbounds a_orth1_2 = conj(a12) * orthogonal1[1] + a22 * orthogonal1[2] + a23 * orthogonal1[3]
-    @inbounds a_orth1_3 = conj(a13) * orthogonal1[1] + conj(a23) * orthogonal1[2] + a33 * orthogonal1[3]
-
-    @inbounds a_orth2_1 = a11 * orthogonal2[1] + a12 * orthogonal2[2] + a13 * orthogonal2[3]
-    @inbounds a_orth2_2 = conj(a12) * orthogonal2[1] + a22 * orthogonal2[2] + a23 * orthogonal2[3]
-    @inbounds a_orth2_3 = conj(a13) * orthogonal2[1] + conj(a23) * orthogonal2[2] + a33 * orthogonal2[3]
-
-    @inbounds c11 = conj(orthogonal1[1])*a_orth1_1 + conj(orthogonal1[2])*a_orth1_2 + conj(orthogonal1[3])*a_orth1_3 - eig2
-    @inbounds c12 = conj(orthogonal1[1])*a_orth2_1 + conj(orthogonal1[2])*a_orth2_2 + conj(orthogonal1[3])*a_orth2_3
-    @inbounds c22 = conj(orthogonal2[1])*a_orth2_1 + conj(orthogonal2[2])*a_orth2_2 + conj(orthogonal2[3])*a_orth2_3 - eig2
-
-    # Solve this robustly (some values might be small or zero)
-    c11² = abs2(c11)
-    c12² = abs2(c12)
-    c22² = abs2(c22)
-    if c11² >= c22²
-        if c11² > 0 || c12² > 0
-            if c11² >= c12²
-                tmp = c12 / c11 # TODO check for compex input
-                p2 = inv(sqrt(1 + abs2(tmp)))
-                p1 = tmp * p2
-            else
-                tmp = c11 / c12 # TODO check for compex input
-                p1 = inv(sqrt(1 + abs2(tmp)))
-                p2 = tmp * p1
+
+    update0(Q, c, s) = Q * @SMatrix [c 0 -s; s 0 c; 0 1 0]
+    update1(Q, c, s) = Q * @SMatrix [0 1 0; c 0 s; -s 0 c]
+    update2(Q, c, s) = Q * @SMatrix [c s 0; s -c 0; 0 0 1]
+    update3(Q, c, s) = Q * @SMatrix [1 0 0; 0 c s; 0 s -c]
+
+    is_rotation = false
+    aggressive = true
+
+    a00, a01, a02, a11, a12, a22 = A[1,1], A[1,2], A[1,3], A[2,2], A[2,3], A[3,3]
+
+    c,s = get_cos_sin(a12, -a02)
+
+    Q = @SMatrix [c s 0; s -c 0; 0 0 1]
+
+    term0 = c * a00 + s * a01
+    term1 = c * a01 + s * a11
+    b00 = c * term0 + s * term1
+    b01 = s * term0 - c * term1
+    term0 = s * a00 - c * a01
+    term1 = s * a01 - c * a11
+    b11 = s * term0 - c * term1
+    b12 = s * a02 - c * a12
+    b22 = a22
+
+    max_iteration = 2 * (1 + 53 + 1021) # 2 * (1 + mantissa_bits - minimum exponent) for Float64
```

Comment:
You can use something like `precision(eltype(A)) - exponent(floatmin(eltype(A)))` to make it more generic.

Reply:
Thanks, implemented as 48fff92.